sql/moz-fx-data-shared-prod/monitoring/shredder_progress/view.sql

```diff
@@ -65,84 +58,136 @@ jobs AS (
+    airflow_task_id,
+    end_date
 ),
 progress AS (
   SELECT
+    airflow_task_id,
+    end_date,
+    LOGICAL_AND(complete) AS complete,
     SUM(slot_ms) AS slot_ms_complete,
-    SUM(slot_ms) + SUM(
-      -- estimate time remaining using bytes_per_slot_ms per table where available
-      (bytes_total - IFNULL(bytes_complete, 0)) / IFNULL(bytes_per_slot_ms, avg_bytes_per_slot_ms)
-    ) AS slot_ms_total,
-    -- tasks start 1 day after end_date
+    -- estimate slot_ms remaining using bytes_per_slot_ms per table where available
+    SUM(bytes_total / IFNULL(bytes_per_slot_ms, avg_bytes_per_slot_ms)) AS slot_ms_total,
     -- diff in seconds to include fractional days
+    MIN(start_time) AS start_time,
     TIMESTAMP_DIFF(
-      current_timestamp,
-      TIMESTAMP(DATE_ADD(MAX(end_date), INTERVAL 1 DAY)),
-      MINUTE
-    ) / 60 / 24 AS days_complete,
+      COALESCE(
+        MAX(end_time),
+        -- start_time of the next airflow run when complete
+        ANY_VALUE(MIN(start_time)) OVER (
+          PARTITION BY
+            airflow_task_id
+          ORDER BY
+            end_date
+          ROWS BETWEEN
+            1 FOLLOWING
+            AND 1 FOLLOWING
+        ),
+        -- current time for the last airflow run
+        CURRENT_TIMESTAMP
+      ),
+      MIN(start_time),
+      SECOND
+    ) AS seconds_complete,
     SUM(bytes_complete) AS bytes_complete,
     SUM(bytes_total) AS bytes_total,
     SUM(tasks_complete) AS tasks_complete,
     SUM(tasks_total) AS tasks_total,
   FROM
     progress_by_target
-  CROSS JOIN
-    avg_bytes_per_slot_ms
+  LEFT JOIN
+    avg_bytes_per_slot_ms AS _
+  USING
+    (airflow_task_id, end_date)
+  GROUP BY
+    airflow_task_id,
+    end_date
 )
 SELECT
-  slot_ms_complete / slot_ms_total * 100 AS days_percent_complete,
-  bytes_complete / bytes_total * 100 AS bytes_percent_complete,
-  tasks_complete / tasks_total * 100 AS tasks_percent_complete,
-  slot_ms_total / slot_ms_complete * days_complete AS days_total,
-  bytes_total / 1024 / 1024 / 1024 / 1024 / 1024 AS petabytes_total,
+  STRUCT(airflow_task_id AS task_id, end_date) AS airflow,
+  start_time,
+  complete,
+  IF(
+    complete,
+    STRUCT(
+      TIMESTAMP_ADD(start_time, INTERVAL seconds_complete SECOND) AS actual,
+      NULL AS estimate_by_slot_ms,
+      NULL AS estimate_by_bytes,
+      NULL AS estimate_by_tasks
+    ),
+    STRUCT(
+      NULL AS actual,
+      TIMESTAMP_ADD(
+        start_time,
+        INTERVAL CAST(
+          IFNULL(slot_ms_total / slot_ms_complete, 1) * seconds_complete AS INT64
+        ) SECOND
+      ) AS estimate_by_slot_ms,
+      TIMESTAMP_ADD(
+        start_time,
+        INTERVAL CAST(bytes_total / bytes_complete * seconds_complete AS INT64) SECOND
+      ) AS estimate_by_bytes,
+      TIMESTAMP_ADD(
+        start_time,
+        INTERVAL CAST(tasks_total / tasks_complete * seconds_complete AS INT64) SECOND
+      ) AS estimate_by_tasks
+    )
+  ) AS completion_time,
+  slot_ms_total,
```

Comment:
I ran this and a large proportion of these rows had a null `slot_ms_total` field. Is this expected?

Reply:
yeah, this only shows slot_ms_total from the last 28 days.